src/Service.Tests/Unittests/SerializationDeserializationTests.cs

```diff
@@ -0,0 +1,403 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.DataApiBuilder.Config.DatabasePrimitives;
+using Azure.DataApiBuilder.Config.ObjectModel;
+using Azure.DataApiBuilder.Core.Services.MetadataProviders.Converters;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Azure.DataApiBuilder.Service.Tests.Unittests
+{
+    [TestClass]
+    [TestCategory("Serialization and Deserialization using SqlMetadataProvider converters")]
+    public class SerializationDeserializationTests
+    {
+        private DatabaseTable _databaseTable;
+        private DatabaseView _databaseView;
+        private DatabaseStoredProcedure _databaseStoredProcedure;
+        private ColumnDefinition _columnDefinition;
+        private ParameterDefinition _parameterDefinition;
+        private SourceDefinition _sourceDefinition;
+        private JsonSerializerOptions _options;
+
+        /// <summary>
+        /// Validates serialization and deserilization of DatabaseTable object
+        /// This test first creates a DatabaseTable object, serializes and deserializes it using our converters
+        /// and verifies that the deserialized object is same as original.
+        /// For DatabaseTable we are checking properties : SchemaName, Name, FullName, SourceType, SourceDefinition and TableDefinition
+        /// This test also tests number of properties currently on DatabaseTable object, this is for future scenario, if there is a new
+        /// property added this test will catch it and also will expose developer to the serialization and deserialization logic
+        /// </summary>
+        [TestMethod]
+        public void TestDatabaseTableSerializationDeserialization()
+        {
+            InitializeObjects();
+
+            // Test to catch if there is change in number of properties/fields
+            // Note: On Addition of property make sure it is added in following object creation _databaseTable and include in serialization
+            // and deserialization test.
+            int fields = typeof(DatabaseTable).GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance).Length;
+            Assert.AreEqual(fields, 6);
+
+            string serializedDatabaseTable = JsonSerializer.Serialize(_databaseTable, _options);
+            DatabaseTable deserializedDatabaseTable = JsonSerializer.Deserialize<DatabaseTable>(serializedDatabaseTable, _options)!;
+
+            Assert.AreEqual(deserializedDatabaseTable.SourceType, _databaseTable.SourceType);
+            Assert.AreEqual(deserializedDatabaseTable.FullName, _databaseTable.FullName);
+            deserializedDatabaseTable.Equals(_databaseTable);
```

Comment:
Are we able to leverage reflection to compare properties?
Relying on developer to update Equals on new props sounds unreliable.

Reply:
yes I tried, but in reflection also we need to use equals at one point. and we cannot equals on all objects. Hence I use reflections atleast to catch the property change so that people are diverted to this test.

Reply:
as discussed offline, will come back as improvement for next round